network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -0,0 +1,134 @@
+package validator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/rs/zerolog"
+
+	channels "github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/network"
+
+	"github.com/onflow/flow-go/model/flow"
+	cborcodec "github.com/onflow/flow-go/network/codec/cbor"
+	"github.com/onflow/flow-go/network/message"
+)
+
+// AuthorizedSenderValidator using the getIdentity func will check if the role of the sender
+// is part of the authorized roles list for the channel being communicated on. A node is considered
+// to be authorized to send a message if all of the following are true.
+// 1. The node is staked
+// 2. The message type is a known message type (can be decoded with cbor codec).
+// 3. The authorized roles list for the channel contains the senders role.
+// 4. The node is not ejected
+func AuthorizedSenderValidator(log zerolog.Logger, channel network.Channel, getIdentity func(peer.ID) (*flow.Identity, bool)) MessageValidator {
+	log = log.With().
+		Str("component", "authorized_sender_validator").
+		Str("network_channel", channel.String()).
+		Logger()
+
+	// use cbor codec to add explicit dependency on cbor encoded messages adding the message type
+	// to the first byte of the message payload, this adds safety against changing codec without updating this validator
+	codec := cborcodec.NewCodec()
```

Comment:
could we have the codec be a parameter?

Reply:
We could add the message type as the first byte regardless of what the codec is

Reply:
My understanding is that currently, the logic of prepending a single-byte "code" to an encoded message has been put directly into the cbor codec implementation: https://github.com/onflow/flow-go/blob/ce0c5aa063f8cbcfb0de058a932dfcc93d0bb5a3/network/codec/cbor/codec.go#L49

However, there's no reason this code-prepending should be specific to Cbor. Instead, we should ideally refactor the code so that we have a "meta-codec" which can take in any arbitrary codec, and encodes messages by prepending a single-byte code to the encoding provided by the underlying codec.

Reply:
I'm planning to refactor the codec parts in a future PR for the unicast case. Essentially, I want to break the message codes out to a separate `MessageType` module, and add a method to the codec interface that returns the `MessageType` for a given message. That should abstract away the details about the specific encoding and how that's represented within the message.

Reply:
@peterargue is there a ticket for the codec refactoring ?

Reply:
We can rearrange the plumbing to make codec a parameter, this will make this PR too big. Best to do all the codec work at once.